src/tests/e2e_extended_mode/src/test.rs

```diff
@@ -374,4 +385,159 @@ impl TestSuite {
+
+        let query_handle = tokio::spawn(async move {
+            client.query("select * from t", &[]).await.unwrap();
+        });
+
+        select! {
+            _ = query_handle => {
+                tracing::error!("Failed to cancel query")
+            },
+            _ = cancel_token.cancel_query(NoTls) => {
+                tracing::trace!("Cancel query successfully")
+            },
+        }
+
+        let new_client = self.create_client(is_distributed).await?;
+
+        let rows = new_client
+            .query("select * from t order by id limit 10", &[])
+            .await?;
+
+        test_eq!(rows.len(), 10);
+        for (expect_id, row) in rows.iter().enumerate() {
+            let id: i32 = row.get(0);
+            test_eq!(id, expect_id as i32);
+        }
+
+        new_client.execute("drop table t", &[]).await?;
+
+        Ok(())
+    }
+
+    async fn complex_cancel(&self, is_distributed: bool) -> anyhow::Result<()> {
+        let client = self.create_client(is_distributed).await?;
+
+        client
+            .execute("create table t1(name varchar, id int)", &[])
+            .await?;
+        client
+            .execute("create table t2(name varchar, id int)", &[])
+            .await?;
+        client
+            .execute("create table t3(name varchar, id int)", &[])
+            .await?;
+
+        let insert_statement = client
+            .prepare_typed("insert INTO t1 (name, id) VALUES ($1, $2)", &[])
+            .await?;
+        let insert_statement2 = client
+            .prepare_typed("insert INTO t2 (name, id) VALUES ($1, $2)", &[])
+            .await?;
+        let insert_statement3 = client
+            .prepare_typed("insert INTO t3 (name, id) VALUES ($1, $2)", &[])
+            .await?;
+        for i in 0..1000 {
+            client
+                .execute(&insert_statement, &[&i.to_string(), &i])
+                .await?;
+            client
+                .execute(&insert_statement2, &[&i.to_string(), &i])
+                .await?;
+            client
+                .execute(&insert_statement3, &[&i.to_string(), &i])
+                .await?;
+        }
+
+        client.execute("flush", &[]).await?;
+
+        client.execute("set query_mode=local", &[]).await?;
+
+        let cancel_token = client.cancel_token();
+
+        let query_sql = "SELECT t1.name, t2.id, t3.name
+        FROM t1
+        INNER JOIN (
+          SELECT id, name
+          FROM t2
+          WHERE id IN (
+            SELECT id
+            FROM t1
+            WHERE name LIKE '%1%'
+          )
+        ) AS t2 ON t1.id = t2.id
+        LEFT JOIN t3 ON t2.name = t3.name
+        WHERE t3.id IN (
+          SELECT MAX(id)
+          FROM t3
+          GROUP BY name
+        )
+        ORDER BY t1.name ASC, t3.id DESC
+        ";
+
+        let query_handle = tokio::spawn(async move {
+            client.query(query_sql, &[]).await.unwrap();
+        });
+
+        select! {
+            _ = query_handle => {
```

Comment:
I test locally and the `cancel_token.cancel_query(NoTls)` always can run successfully. 

But there may be a case that the cancel_query will not be scheduled in deterministic test🤔